### ISSUE TITLE
git hooks using wrong Ruby when hub aliased to git

### ISSUE THREAD

**Issue by deivid-rodriguez** [github/hub#855]
To reproduce

    git clone deivid-rodriguez/byebug
    cd byebug
    gem install bundler
    bundle install
    bundle exec overcommit --install
    sed -i '/enabled: false/d' .overcommit.yml
    sed -i '/fast_finish/d' .travis.yml
    git add .travis.yml
    git commit

And I get

    Running pre-commit hooks
    Checking for trailing whitespace.................[TrailingWhitespace] OK
    Checking Travis CI configuration.........................[TravisLint] FAILED
    /home/davidr/.gem/ruby/2.2.1/gems/bundler-1.9.1/lib/bundler/spec_set.rb:92:in `block in materialize': Could not find rake-10.4.2 in any of the sources (Bundler::GemNotFound)
        from /home/davidr/.gem/ruby/2.2.1/gems/bundler-1.9.1/lib/bundler/spec_set.rb:85:in `map!'
        from /home/davidr/.gem/ruby/2.2.1/gems/bundler-1.9.1/lib/bundler/spec_set.rb:85:in `materialize'
        from /home/davidr/.gem/ruby/2.2.1/gems/bundler-1.9.1/lib/bundler/definition.rb:132:in `specs'
        from /home/davidr/.gem/ruby/2.2.1/gems/bundler-1.9.1/lib/bundler/definition.rb:177:in `specs_for'
        from /home/davidr/.gem/ruby/2.2.1/gems/bundler-1.9.1/lib/bundler/definition.rb:166:in `requested_specs'
        from /home/davidr/.gem/ruby/2.2.1/gems/bundler-1.9.1/lib/bundler/environment.rb:18:in `requested_specs'
        from /home/davidr/.gem/ruby/2.2.1/gems/bundler-1.9.1/lib/bundler/runtime.rb:13:in `setup'
        from /home/davidr/.gem/ruby/2.2.1/gems/bundler-1.9.1/lib/bundler.rb:122:in `setup'
        from /home/davidr/.gem/ruby/2.2.1/gems/bundler-1.9.1/lib/bundler/setup.rb:18:in `<top (required)>'
        from /usr/lib/ruby/1.9.1/rubygems/custom_require.rb:36:in `require'
        from /usr/lib/ruby/1.9.1/rubygems/custom_require.rb:36:in `require'
        Checking YAML syntax.....................................[YamlSyntax] OK

If I disable the `hub` -> `git` alias, it works fine.

I've reproduced this with both `chruby` and `rvm`.

Thanks!!

Originally [reported to overcommit](https://github.com/causes/overcommit/issues/153) but I don't think the issue belongs there.

**Comment by mislav** [comment]
How did you install hub? Check its shebang:

```
file `which hub` | grep text && head -1 `which hub`
```

**Comment by deivid-rodriguez** [comment]
I installed it with `brew install hub`, but with `brew` provided by [linuxbrew](https://github.com/Homebrew/linuxbrew). This is the output of your command.

    $ file `which hub` | grep text && head -1 `which hub`
    $ file `which hub`
    /home/davidr/.linuxbrew/bin/hub: symbolic link to `../Cellar/hub/2.2.0/bin/hub' 
    $ file ~/.linuxbrew/Cellar/hub/2.2.0/bin/hub | grep text
    $ file ~/.linuxbrew/Cellar/hub/2.2.0/bin/hub 
    /home/davidr/.linuxbrew/Cellar/hub/2.2.0/bin/hub: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), not stripped

If I uninstall it and reinstall using the https://github.com/github/hub#standalone method, I get basically the same thing.

    $ file `which hub` | grep text && head -1 `which hub`
    $ file `which hub`
    /usr/local/bin/hub: ELF 64-bit LSB  executable, x86-64, version 1 (SYSV), dynamically linked (uses shared libs), not stripped

Thanks for your help!

**Comment by mislav** [comment]
OK, so it's latest Go hub on Linux. Good to know.

Which Ruby version did you expect to get activated instead of 2.2.1? Are you using `chruby` or `rvm`, or both at the same time (not sure if that's even possible)? When using `chruby`, I trust that the desired version of Ruby is pushed onto PATH automatically when you enter a directory, right? Are you saying that running git hooks through hub (aliased as `git`) somehow manage to pick a version that is not on current PATH?

**Comment by deivid-rodriguez** [comment]
> Which Ruby version did you expect to get activated instead of 2.2.1?

I actually expected 2.2.1 but it seems system's Ruby (1.9), look at the first two lines of the backtrace.

> Are you using chruby or rvm, or both at the same time (not sure if that's even possible)?

I was using `rvm`. Then I removed `rvm` and installed `chruby` to see if it was related to the version manager but I got the same results.

> When using chruby, I trust that the desired version of Ruby is pushed onto PATH automatically when you enter a directory, right? Are you saying that running git hooks through hub (aliased as git) somehow manage to pick a version that is not on current PATH?

Not sure how `chruby` works internally...

Can you reproduce the bug yourself of this is something with my environment?

Thanks.

**Comment by deivid-rodriguez** [comment]
@mislav Not sure if it's relevant but this happens with all hooks `overcommit `installs, which is actually a single one (all hook types are symlinked to it)  and starts with the shebang:

    #!/usr/bin/env ruby

**Comment by mislav** [comment]
> Can you reproduce the bug yourself of this is something with my environment?

I will try.

**Comment by deivid-rodriguez** [comment]
Thanks! I found an easier reproduction (independent from `overcommit`). The following assumes you're currently not using 1.9.3 and your system Ruby is 1.9.3

    $ mkdir hub_issue_855
    $ cd hub_issue_855
    $ git init
    $ echo '#!/usr/bin/env ruby' > .git/hooks/pre-commit
    $ echo 'fail("Oops") if RUBY_VERSION == "1.9.3"' >> .git/hooks/pre-commit
    $ chmod +x .git/hooks/pre-commit
    $ echo "First readme" > README.txt
    $ git add .
    $ git commit
    .git/hooks/pre-commit:2:in `<main>': Oops (RuntimeError)

**Comment by deivid-rodriguez** [comment]
@mislav Did you have time to have a look or reproduce this? Thanks!

**Comment by mislav** [comment]
I can't reproduce using rbenv. Git hooks behave the same regardless of whether `git` is running `hub` or plain `git`.

**Comment by mislav** [comment]
Can't reproduce using latest chruby either. In order to support RVM in hooks, you can configure their shebang to run them through an RVM wrapper script. I have **no idea** why you get different results going through `hub`.

**Comment by deivid-rodriguez** [comment]
Me neither... :disappointed: Did you try it in linux?

I'll try to setup a fresh VM and see what happens. I'll post my results when I do.

**Comment by deivid-rodriguez** [comment]
Hi again @mislav, I got some time to investigate this.

This is certainly not a bug in `hub`, but a bug in `git` in my opinion. The problem is that, when called with an absolute path (`/usr/bin/git`),  `git` prepends `/usr/bin` to the path. I can't think of any good reason why `git` would do that and I'm assuming it's not intentional. So I'll report it as an issue to `git` and see what they say.

`hub` actually calls `git` with an absolute path because `syscall.exec` requires an absolute path to be passed in. In Windows systems, `hub` uses `spawn` when delegating execution to `git`, so there the bug is not reproduced. I actually fixed my problem by forcing `hub` to use always `spawn`:

```diff
diff --git a/cmd/cmd.go b/cmd/cmd.go
index 55e7b8c..ebe229b 100644
--- a/cmd/cmd.go
+++ b/cmd/cmd.go
@@ -4,9 +4,7 @@ import (
        "fmt"
        "os"
        "os/exec"
-       "runtime"
        "strings"
-       "syscall"
 
        "github.com/github/hub/Godeps/_workspace/src/github.com/kballard/go-shellquote"
        "github.com/github/hub/utils"
@@ -41,16 +39,6 @@ func (cmd *Cmd) CombinedOutput() (string, error) {
        return string(output), err
 }
 
-// Run runs command with `Exec` on platforms except Windows
-// which only supports `Spawn`
-func (cmd *Cmd) Run() error {
-       if runtime.GOOS == "windows" {
-               return cmd.Spawn()
-       } else {
-               return cmd.Exec()
-       }
-}
-
 // Spawn runs command with spawn(3)
 func (cmd *Cmd) Spawn() error {
        c := exec.Command(cmd.Name, cmd.Args...)
@@ -61,20 +49,6 @@ func (cmd *Cmd) Spawn() error {
        return c.Run()
 }
 
-// Exec runs command with exec(3)
-// Note that Windows doesn't support exec(3): http://golang.org/src/pkg/syscall/exec_windows.go#L339
-func (cmd *Cmd) Exec() error {
-       binary, err := exec.LookPath(cmd.Name)
-       if err != nil {
-               return fmt.Errorf("command not found: %s", cmd.Name)
-       }
-
-       args := []string{binary}
-       args = append(args, cmd.Args...)
-
-       return syscall.Exec(binary, args, os.Environ())
-}
-
 func New(cmd string) *Cmd {
        cmds, err := shellquote.Split(cmd)
        utils.Check(err)
diff --git a/commands/runner.go b/commands/runner.go
index f449413..27f8c4d 100644
--- a/commands/runner.go
+++ b/commands/runner.go
@@ -112,14 +112,9 @@ func printCommands(cmds []*cmd.Cmd) {
 }
 
 func executeCommands(cmds []*cmd.Cmd) error {
-       for i, c := range cmds {
+       for _, c := range cmds {
                var err error
-               // Run with `Exec` for the last command in chain
-               if i == len(cmds)-1 {
-                       err = c.Run()
-               } else {
-                       err = c.Spawn()
-               }
+               err = c.Spawn()
 
                if err != nil {
                        return err
diff --git a/git/git.go b/git/git.go
index 51ebe3d..7e52ffd 100644
--- a/git/git.go
+++ b/git/git.go
@@ -206,7 +206,7 @@ func Run(command string, args ...string) error {
                cmd.WithArg(a)
        }
 
-       return cmd.Run()
+       return cmd.Spawn()
 }
 
 func gitOutput(input ...string) (outputs []string, err error) {
```

This has been causing issues with git hooks implemented in Ruby for a while. See https://github.com/sstephenson/rbenv/issues/374, for example.

**Comment by mislav** [comment]
Thanks for the diff above about changing Exec to Spawn, but I don't get how it fixes things. How does `/usr/bin/git` know whether it has been called with an absolute path or not?

**Comment by jingweno** [comment]
The use of `Exec` instead of `Spawn` where possible is to fix pagination issue in #708.

**Comment by deivid-rodriguez** [comment]
> Thanks for the diff above about changing Exec to Spawn, but I don't get how it fixes things. How does /usr/bin/git know whether it has been called with an absolute path or not?

You are welcome. I just added the patch for reference, it's not meant to be merged. I'm sure (and @jingweno confirmed) there's good reasons to prefer `Exec` when available. 

I guess `git` parses `argv[0]` to know that, but reading the source code or asking some familiarized with it will better answer that question.

As I said, I'll open an issue in `git` and get back here when I have answers.

**Comment by deivid-rodriguez** [comment]
It seems that this will be fixed in `git` in the next major version. Closing.

**Comment by mislav** [comment]
Thanks for the investigation!

**Comment by deivid-rodriguez** [comment]
No problem! :smiley:

**Comment by deivid-rodriguez** [comment]
For completeness, the fix for this [has been merged into git's master](https://github.com/git/git/commit/1645dbeff75d466230eb54b8548cfe9c9dcc9e5e) and, I guess, will be released with `git 2.5.0`.
